go/api/callbacks.go

```diff
@@ -0,0 +1,114 @@
+// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
+
+package api
+
+// #include "../lib.h"
+import "C"
+
+import (
+	"fmt"
+	"sync"
+	"unsafe"
+
+	"github.com/valkey-io/valkey-glide/go/api/errors"
+)
+
+// Registry to track clients by their pointer address
+var (
+	clientRegistry   = make(map[uintptr]*baseClient)
+	clientRegistryMu sync.RWMutex
+)
+
+// RegisterClient registers a client in the registry using its pointer value
+func RegisterClient(client *baseClient, ptrValue uintptr) {
+	clientRegistryMu.Lock()
+	defer clientRegistryMu.Unlock()
+	clientRegistry[ptrValue] = client
+}
+
+// UnregisterClient removes a client from the registry
+func UnregisterClient(ptrValue uintptr) {
+	clientRegistryMu.Lock()
+	defer clientRegistryMu.Unlock()
+	delete(clientRegistry, ptrValue)
+}
+
+// GetClientByPtr gets a client from the registry by its pointer value
+func GetClientByPtr(ptrValue uintptr) *baseClient {
+	clientRegistryMu.RLock()
+	defer clientRegistryMu.RUnlock()
+	return clientRegistry[ptrValue]
+}
+
+//export successCallback
+func successCallback(channelPtr unsafe.Pointer, cResponse *C.struct_CommandResponse) {
+	response := cResponse
+	resultChannel := *(*chan payload)(getPinnedPtr(channelPtr))
+	resultChannel <- payload{value: response, error: nil}
+}
+
+//export failureCallback
+func failureCallback(channelPtr unsafe.Pointer, cErrorMessage *C.char, cErrorType C.RequestErrorType) {
+	defer C.free_error_message(cErrorMessage)
+	msg := C.GoString(cErrorMessage)
+	resultChannel := *(*chan payload)(getPinnedPtr(channelPtr))
+	resultChannel <- payload{value: nil, error: errors.GoError(uint32(cErrorType), msg)}
+}
+
+//export pubSubCallback
+func pubSubCallback(clientPtr unsafe.Pointer, pushKind C.PushKind, message unsafe.Pointer, message_len C.int, channel unsafe.Pointer, channel_len C.int, pattern unsafe.Pointer, pattern_len C.int) {
```

Comment:
PushKind is not used actually. The single purpose of that was to filter messages - you already did that in rust code.
```suggestion
func pubSubCallback(clientPtr unsafe.Pointer, message unsafe.Pointer, message_len C.int, channel unsafe.Pointer, channel_len C.int, pattern unsafe.Pointer, pattern_len C.int) {
```

Or you can keep it for future

Reply:
PushKind is used for getting the PubSubMessage object depending on whether it is a PushMessage, PushSMessage or PushPMessage.